### PR TITLE
[RLlib] Add torch models

### PR DESCRIPTION
**Pull request by smorad**
## Why are these changes needed?
This adds a pytorch implementation of the new abstract base models. Please see the prerequisite before approving/merging this commit.
## Related issue number
Prerequisite:  https://github.com/ray-project/ray/pull/29009/files

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by kouroshHakha — COMMENTED** [review]
Thanks, look at the comments. Need more work on the documentation part.

File: rllib/models/torch/model.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    If implementing a custom recurrent model, you likely want to inherit
+    from this model.
+
+    Args:
+        config: The config used to construct the model
+
+    """
+
+    def __init__(self, config: ModelConfig) -> None:
+        Model.__init__(self)
```
[the remark this answers is not in the document]
Comment (smorad) [reply rllib/models/torch/model.py:55]: Nice catch

File: rllib/models/torch/model.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+from ray.rllib.utils.annotations import (
+    OverrideToImplementCustomLogic,
+    PublicAPI,
+)
+from ray.rllib.models.temp_spec_classes import TensorDict, ModelConfig
+from ray.rllib.models.base_model import RecurrentModel, Model, ModelIO
+
+
+class TorchModelIO(ModelIO):
```
[the remark this answers is not in the document]
Comment (smorad) [reply rllib/models/torch/model.py:13]: Hmm. I don't like the triple inheritance thing we currently have, but I think sticking the `nn.Module` in `TorchModelIO` might be less clear. `nn.Module` is integral for a module, and there might be some cases where the user doesn't want to inherit `IOModule`. Like if they JIT the model or something. IDK perhaps we can discuss this further during our meeting.

File: rllib/models/torch/model.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    def load(self, path: str) -> RecurrentModel:
+        """Loads the state dict from the specified path
+
+        Args:
+            path: Path on disk to load the checkpoint from
+        """
+        self.load_state_dict(torch.load(path))
+
+
+class TorchRecurrentModel(RecurrentModel, nn.Module, TorchModelIO):
```
[the remark this answers is not in the document]
Comment (smorad) [reply rllib/models/torch/model.py:43]: Is it worth doing something like

```python
class TorchRecurrentModel(...):
    def _unroll(...):
        return super().unroll(...)
```

To make it clear that these functions exist? Or should I just write in the docstring?

File: rllib/models/torch/model.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kouroshHakha) [reply rllib/models/torch/model.py:13]: ok, convinced :)

File: rllib/models/torch/model.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+        # Save model weights to path
+        save(self, path: str) -> None
+
+        # Load model weights from path
+        load(self, path: str) -> None
+
+    Examples:
+        >>> class MyCustomModel(TorchRecurrentModel):
+        ...     def __init__(self, input_size, output_size, recurrent_size):
+        ...         self.input_spec = SpecDict(
```
Comment (kouroshHakha) [inline rllib/models/torch/model.py:80]: respect the getters :)

File: rllib/models/torch/model.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+    Required Overrides:
+        # Define unrolling (forward pass) over a sequence of inputs
+        _forward(self, inputs: TensorDict, **kwargs)
+            -> TensorDict
+
+    Optional Overrides:
+        # Save model weights to path
+        save(self, path: str) -> None
+
+        # Load model weights from path
```
Comment (kouroshHakha) [inline rllib/models/torch/model.py:153]: _check_input and _check_output is also optional override

File: rllib/models/torch/model.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+        load(self, path: str) -> None
+
+    Examples:
+        >>> class MyCustomModel(TorchModel):
+        ...     def __init__(self, input_size, hidden_size, output_size):
+        ...         self.input_spec = SpecDict(
```
Comment (kouroshHakha) [inline rllib/models/torch/model.py:159]: same problem here.

File: rllib/models/torch/model.py (hunk 4 above)
Comment (kouroshHakha) [reply rllib/models/torch/model.py:80]: also use super(). Make sure in the docs you mention super should be called for init.

Comment (kouroshHakha) [reply rllib/models/torch/model.py:80]: also for input take config as input. Don't break them up here. We'll figure out the data type for config objects later. For now assume they are dictionaries.

**Review by kouroshHakha — COMMENTED** [review]
left some comments. WDYT?

File: rllib/models/torch/model.py (hunk 4 above)
Comment (smorad) [reply rllib/models/torch/model.py:80]: fixed

File: rllib/models/torch/model.py (hunk 5 above)
Comment (smorad) [reply rllib/models/torch/model.py:153]: fixed

File: rllib/models/torch/model.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    """
+
+    def __init__(self, config: ModelConfig) -> None:
+        Model.__init__(self)
+        nn.Module.__init__(self)
+        TorchModelIO.__init__(self, config)
+        self._config = config
+
+    @OverrideToImplementCustomLogic
+    def _initial_state(self) -> TensorDict:
```
[the remark this answers is not in the document]
Comment (smorad) [reply rllib/models/torch/model.py:113]: fixed

File: rllib/models/torch/model.py (hunk 6 above)
Comment (smorad) [reply rllib/models/torch/model.py:159]: fixed

**Comment by kouroshHakha** [comment]
@smorad Lint is failing.

**Comment by kouroshHakha** [comment]
> @smorad Lint is failing.

ok everything else is failing :)

File: rllib/models/torch/model.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (gjoliver) [reply rllib/models/torch/model.py:43]: start with just doc string?

**Review by gjoliver — APPROVED** [review]
this looks pretty nice.

**Comment by gjoliver** [comment]
Lint error is unrelated. we will monitor the agent connectors test to make sure nothing is actually broken (should not be related.).

**Pull request merged**